### ISSUE TITLE
Storage Backend: OpenStack Object Storage (Swift)

### ISSUE THREAD

**Issue by ar-jan** [restic/restic#272]
Could OpenStack Object Storage be supported by restic? For example RunAbove object storage looks interesting, since it's cheaper than Amazon S3, and especially outgoing traffic is much cheaper.

**Comment by fd0** [comment]
According to https://www.openstack.org/software/openstack-storage/, it already supports the S3 API, so it should work with restic.

**Comment by ar-jan** [comment]
Ah, excellent!

**Comment by ar-jan** [comment]
It seems this needs some more work and/or that it would be useful to support native Swift. Apparently OpenStack providers can choose not to enable S3 emulation, [like RunAbove](https://community.runabove.com/share/topic/s3-emulation-enabled-on-storage).

I'm also not sure if they did support S3 emulation how exactly one would use it currently - would you substitute AWS_ACCESS_KEY_ID with the Swift username (email address), and AWS_SECRET_ACCESS_KEY with the password? I suppose there'd also need to be a way to provide other service endpoints than AWS regions.

**Comment by fd0** [comment]
did you try to use restic with the s3 backend with swift? did it work?

**Comment by ar-jan** [comment]
I tried, though all RunAbove examples use Swift, and require other environment variables than S3, e.g. Tenant name ([example](https://community.runabove.com/kb/en/object-storage/how-to-store-your-backup-with-duplicity-and-openstack-swift.html)). I tried filling the AWS environment variables with RunAbove username and password, and using  s3://auth.runabove.io/v2.0/container_name, and got "unsupported protocol scheme". But I guess that's expected, since restic's S3 backend interprets the this URL as an Amazon S3 region name.

How could one currently specify an s3:// backend with a non-Amazon endpoint?

**Comment by fd0** [comment]
Ok, this needs to be investigated. I'll reopen the issue for now.

**Comment by retrack** [comment]
Hi there,

Tested with non-Amazon endpoint at Exoscale ( S3 storage backed with Pithos.io)

It fails with the following message at `init`stage:

    unsupported protocol scheme

Using debug mode did not provide more information.

**Comment by fd0** [comment]
This is a limitiation of the s3 backend code (you cannot specify a server other than Amazon at the moment), and it's tracked in #315

**Comment by byo** [comment]
Hi,

I was able to create a base implementation of native swift support which you can find here: https://github.com/byo/restic/tree/swift-backend.

Why native? To support all swift-specific options such as keystone-based authentication, swift regions, container policies and swift backends without s3 compatibility. It should also help with hubic integration (#536) which is also built on top of swift, I haven't tested it though.

The implementation is based on the s3 one so the code is pretty similar, I was testing it with OVH's public could storage (https://www.ovh.com/us/cloud/storage/object-storage.xml). Data looks ok, I was able to backup ~80GB of files, currently I'm validating of md5s in fuse-mounted snapshot, so far everything is ok.

There's one test failing for me, it looks like OVH's public cloud storage is doing some kind of delayed delete of object and testing if one exists shortly after being removed could occasionally end up with false positive.

Please let me know what do you think about this.

**Comment by fd0** [comment]
Thanks for giving this a try. What (backend) specific options/authentication tokens or similar specific things does this backend need?

**Comment by byo** [comment]
I'm trying to mimic what's available in official swift client, there's a plenty of options and it depends on the authentication protocol version: http://docs.openstack.org/developer/python-swiftclient/cli.html#authentication. What's supported is legacy and v2 authentication (I tested with v2 since that's what used in OVH), v3 is partially supported (no OS_USER_ID nor OS_PROJECT_ID).

I also support preauth mode where authentication is done externally through keystone server, the caller sets up OS_STORAGE_URL and OS_AUTH_TOKEN.

In addition to authentication, one can specify OS_REGION_NAME because one authentication endpoint and one user identity could be used by many independent swift servers.

I've switched to different swift server recently (a more loaded one) and it looks like there are still some timeout problems to fix. I hope to find some time next week to figure this out.
